src/crypto/ocb_openssl.cc

```diff
@@ -1,1206 +1,162 @@
-{
-	memset(ctx, 0, sizeof(ae_ctx));
-	return AE_SUCCESS;
+int ae_ctx_sizeof(void) {
+  return sizeof(_ae_ctx);
 }
 
-int ae_ctx_sizeof(void) { return (int) sizeof(ae_ctx); }
-
-/* ----------------------------------------------------------------------- */
-
-int ae_init(ae_ctx *ctx, const void *key, int key_len, int nonce_len, int tag_len)
-{
-    unsigned i;
-    block tmp_blk;
-
-    if (nonce_len != 12)
-    	return AE_NOT_SUPPORTED;
-
-    /* Initialize encryption & decryption keys */
-    #if (OCB_KEY_LEN > 0)
-    key_len = OCB_KEY_LEN;
-    #endif
-    AES_set_encrypt_key((unsigned char *)key, key_len*8, &ctx->encrypt_key);
-    AES_set_decrypt_key((unsigned char *)key, (int)(key_len*8), &ctx->decrypt_key);
-
-    /* Zero things that need zeroing */
-    ctx->cached_Top = ctx->ad_checksum = zero_block();
-    ctx->ad_blocks_processed = 0;
-
-    /* Compute key-dependent values */
-    AES_encrypt((unsigned char *)&ctx->cached_Top,
-                            (unsigned char *)&ctx->Lstar, &ctx->encrypt_key);
-    tmp_blk = swap_if_le(ctx->Lstar);
-    tmp_blk = double_block(tmp_blk);
-    ctx->Ldollar = swap_if_le(tmp_blk);
-    tmp_blk = double_block(tmp_blk);
-    ctx->L[0] = swap_if_le(tmp_blk);
-    for (i = 1; i < L_TABLE_SZ; i++) {
-		tmp_blk = double_block(tmp_blk);
-    	ctx->L[i] = swap_if_le(tmp_blk);
-    }
-
-    #if (OCB_TAG_LEN == 0)
-    	ctx->tag_len = tag_len;
-    #else
-    	(void) tag_len;  /* Suppress var not used error */
-    #endif
-
-    return AE_SUCCESS;
+int ae_init(ae_ctx *ctx, const void *key_param, int key_len, int nonce_len,
+	    int tag_len) {
+  // Pre-condition: Only nonces of length 12 are supported. The
+  // documentation of `ae_init` in ae.h specifies that `ctx` is
+  // untouched if an invalid configuration is requested. Delegating
+  // this to OpenSSL would happen too late; `ctx` has already been
+  // modified.
+  if (nonce_len != 12) {
+    return AE_NOT_SUPPORTED;
+  }
+
+  // Pre-condition: Only AES-128 is supported.
+  if (key_len != 16) {
+    return AE_NOT_SUPPORTED;
+  }
+
+  // Create an OpenSSL EVP context. It does not yet have any specific
+  // cipher associated with it.
+  if (!(ctx->ocb_ctx = EVP_CIPHER_CTX_new())) {
+    return -10;
+  }
+  // ae_init is used for both OCB encryption and decryption. However,
+  // the EVP_CIPHER_CTX wants to know if the operation is encryption
+  // or decryption. For now, arbitrary pick encryption. This will get
+  // overriden in the `ae_encrypt` and `ae_decrypt` methods when the
+  // key and nonce (IV) are loaded.
+  if (EVP_EncryptInit_ex(ctx->ocb_ctx, EVP_aes_128_ocb(),
+			 NULL, NULL, NULL) != 1) {
+    return -11;
+  }
+  // Attempt to set the nonce length. If it fails, the length must not
+  // be supported. However, that should have been handled by the
+  // pre-condition check above.
+  if (EVP_CIPHER_CTX_ctrl(ctx->ocb_ctx, EVP_CTRL_AEAD_SET_IVLEN,
+			  nonce_len, NULL)!= 1) {
```

Comment:
I think the default IV length is already 12. Default tag length is 16. That matches what crypto.cc uses, so you could, optionally, skip the SET_IVLEN and SET_TAG ctrl calls. That, in turn, should avoid needing the funny NULL, NULL init step.

Reply:
If I understood the NULL, NULL init step it's imbuing the EVP_CIPHER object with any per-algorithm initialization, and only the key, nonce call later is what personalizes it for the specific keypair. Although the defaults are as you say (and we don't support anything else) I'd prefer to leave the calls in just in case some future version of OpenSSL and OCB-AES (however unlikely) decides to change the API.